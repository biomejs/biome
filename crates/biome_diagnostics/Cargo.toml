[package]
authors.workspace = true
categories        = ["development-tools"]
description       = "Pretty error reporting library"
documentation     = "https://docs.rs/biome_diagnostics"
edition.workspace = true
license.workspace = true
name              = "biome_diagnostics"
version           = "0.2.0"

[[example]]
name = "cli"
test = true

[[example]]
name = "fs"
test = true

[[example]]
name = "lint"
test = true

[[example]]
name = "serde"
test = true

[dependencies]
backtrace                    = "0.3.66"
biome_console                = { workspace = true, features = ["serde_markup"] }
biome_diagnostics_categories = { workspace = true, features = ["serde"] }
biome_diagnostics_macros     = { workspace = true }
biome_rowan                  = { workspace = true }
biome_text_edit              = { workspace = true }
biome_text_size              = { workspace = true }
bitflags                     = { workspace = true }
bpaf                         = { workspace = true }
schemars                     = { workspace = true, optional = true }
serde                        = { workspace = true, features = ["derive"] }
termcolor                    = "1.1.2"
unicode-width                = "0.1.9"

[features]
schema = ["schemars", "biome_text_edit/schemars", "biome_diagnostics_categories/schemars"]

[dev-dependencies]
serde_json = { workspace = true }
trybuild   = "1.0.80"
