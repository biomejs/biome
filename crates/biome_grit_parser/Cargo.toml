[package]
authors.workspace    = true
categories.workspace = true
description          = "Biome's GritQL parser"
edition.workspace    = true
homepage.workspace   = true
keywords.workspace   = true
license.workspace    = true
name                 = "biome_grit_parser"
repository.workspace = true
version              = "0.1.0"

[dependencies]
biome_console      = { workspace = true }
biome_diagnostics  = { workspace = true }
biome_grit_factory = { workspace = true }
biome_grit_syntax  = { workspace = true }
biome_parser       = { workspace = true }
biome_rowan        = { workspace = true }
bitflags           = { workspace = true }
schemars           = { workspace = true, optional = true }
serde              = { workspace = true, features = ["derive"] }
serde_json         = { workspace = true }
smallvec           = { workspace = true }
tracing            = { workspace = true }
unicode-bom        = { workspace = true }

[dev-dependencies]
insta             = { workspace = true }
quickcheck        = { workspace = true }
quickcheck_macros = { workspace = true }
tests_macros      = { workspace = true }

[features]
schemars = ["dep:schemars"]
serde    = ["biome_grit_syntax/serde"]
tests    = []

# cargo-workspaces metadata
[package.metadata.workspaces]
independent = true

[lints]
workspace = true
