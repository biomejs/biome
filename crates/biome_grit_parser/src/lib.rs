mod constants;
mod lexer;
mod parser;
mod token_source;

use biome_grit_factory::GritSyntaxFactory;
use biome_grit_syntax::{GritLanguage, GritRoot, GritSyntaxNode};
use biome_parser::diagnostic::ParseDiagnostic;
use biome_parser::tree_sink::LosslessTreeSink;
use biome_rowan::{AstNode, NodeCache};
use parser::{parse_root, GritParser};

pub(crate) type GritLosslessTreeSink<'source> =
    LosslessTreeSink<'source, GritLanguage, GritSyntaxFactory>;

pub fn parse_grit(source: &str) -> GritParse {
    let mut cache = NodeCache::default();
    parse_grit_with_cache(source, &mut cache)
}

/// Parses the provided string as a GritQL pattern using the provided node cache.
pub fn parse_grit_with_cache(source: &str, cache: &mut NodeCache) -> GritParse {
    tracing::debug_span!("parse").in_scope(move || {
        let mut parser = GritParser::new(source);

        parse_root(&mut parser);

        let (events, diagnostics, trivia) = parser.finish();

        let mut tree_sink = GritLosslessTreeSink::with_cache(source, &trivia, cache);
        biome_parser::event::process(&mut tree_sink, events, diagnostics);
        let (green, diagnostics) = tree_sink.finish();

        GritParse::new(green, diagnostics)
    })
}

/// A utility struct for managing the result of a parser job
#[derive(Debug)]
pub struct GritParse {
    root: GritSyntaxNode,
    diagnostics: Vec<ParseDiagnostic>,
}

impl GritParse {
    pub fn new(root: GritSyntaxNode, diagnostics: Vec<ParseDiagnostic>) -> Self {
        Self { root, diagnostics }
    }

    /// The syntax node represented by this Parse result
    ///
    /// ```
    /// # use biome_grit_parser::parse_grit;
    /// # use biome_grit_syntax::{GritSyntaxKind, AnyGritLiteral, AnyGritPattern, GritRootExt};
    /// # use biome_rowan::{AstNode, AstNodeList, SyntaxError};
    ///
    /// # fn main() {
    /// use biome_grit_syntax::GritSyntaxKind;
    /// let parse = parse_grit(r#"`console.log($message)`"#);
    ///
    /// // Get the pattern
    /// let pattern = parse.tree().pattern();
    /// match pattern {
    ///     Some(AnyGritPattern::AnyGritLiteral(AnyGritLiteral::GritCodeSnippet(snippet))) => {
    ///         assert_eq!(
    ///             snippet.source().unwrap().syntax().kind(),
    ///             GritSyntaxKind::GRIT_BACKTICK_SNIPPET_LITERAL
    ///         );
    ///     }
    ///     _ => panic!("Unexpected pattern"),
    /// }
    /// # }
    /// ```
    pub fn syntax(&self) -> GritSyntaxNode {
        self.root.clone()
    }

    /// Get the diagnostics which occurred when parsing
    pub fn diagnostics(&self) -> &[ParseDiagnostic] {
        &self.diagnostics
    }

    /// Get the diagnostics which occurred when parsing
    pub fn into_diagnostics(self) -> Vec<ParseDiagnostic> {
        self.diagnostics
    }

    /// Returns [true] if the parser encountered some errors during the parsing.
    pub fn has_errors(&self) -> bool {
        self.diagnostics
            .iter()
            .any(|diagnostic| diagnostic.is_error())
    }

    /// Convert this parse result into a typed AST node.
    ///
    /// # Panics
    /// Panics if the node represented by this parse result mismatches.
    pub fn tree(&self) -> GritRoot {
        GritRoot::unwrap_cast(self.syntax())
    }
}
