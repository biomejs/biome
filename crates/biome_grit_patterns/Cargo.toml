[package]
authors.workspace    = true
categories.workspace = true
description          = "Biome implementing for matching Grit Patterns"
edition.workspace    = true
homepage.workspace   = true
keywords.workspace   = true
license.workspace    = true
name                 = "biome_grit_patterns"
publish              = false
repository.workspace = true
version              = "0.0.1"

[dependencies]
anyhow               = { workspace = true }
biome_console        = { workspace = true }
biome_diagnostics    = { workspace = true }
biome_grit_parser    = { workspace = true }
biome_grit_syntax    = { workspace = true }
biome_js_syntax      = { workspace = true }
biome_parser         = { workspace = true }
biome_rowan          = { workspace = true }
grit-pattern-matcher = { version = "0.2" }
grit-util            = { version = "0.2" }
im                   = { version = "15.1.0" }
itertools            = "0.10.5"
serde                = { workspace = true, features = ["derive"] }

[lints]
workspace = true
