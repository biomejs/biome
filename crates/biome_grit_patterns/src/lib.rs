mod errors;
mod parse;
mod pattern;

pub use errors::*;
pub use parse::parse_pattern;
pub use pattern::GritPattern;
