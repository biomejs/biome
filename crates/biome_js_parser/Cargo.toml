[package]
authors.workspace    = true
categories           = ["parser-implementations", "development-tools"]
description          = "An extremely fast ECMAScript parser for Biome, forked from the RSlint project"
documentation        = "https://docs.rs/biome_js_parser"
edition.workspace    = true
homepage.workspace   = true
license.workspace    = true
name                 = "biome_js_parser"
repository.workspace = true
version              = "0.1.0"

[dependencies]
biome_console          = { workspace = true }
biome_diagnostics      = { workspace = true }
biome_js_factory       = { workspace = true }
biome_js_syntax        = { workspace = true }
biome_js_unicode_table = { workspace = true }
biome_parser           = { workspace = true }
biome_rowan            = { workspace = true }
bitflags               = { workspace = true }
cfg-if                 = "1.0.0"
drop_bomb              = "0.1.5"
indexmap               = { workspace = true }
schemars               = { workspace = true, optional = true }
serde                  = { workspace = true, features = ["derive"] }
serde_json             = { workspace = true }
smallvec               = { workspace = true }
tracing                = { workspace = true }

[dev-dependencies]
biome_js_parser   = { workspace = true, features = ["serde"] }
expect-test       = "1.2.2"
quickcheck        = { workspace = true }
quickcheck_macros = { workspace = true }
tests_macros      = { workspace = true }

[features]
schemars = ["dep:schemars"]
serde    = ["biome_js_syntax/serde"]
tests    = []

# cargo-workspaces metadata
[package.metadata.workspaces]
independent = true
