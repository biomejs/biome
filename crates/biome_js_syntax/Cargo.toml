[package]
authors.workspace    = true
description          = "SyntaxKind and common rowan definitions for biome_js_parser"
documentation        = "https://docs.rs/biome_js_syntax"
edition.workspace    = true
license.workspace    = true
name                 = "biome_js_syntax"
repository.workspace = true
version              = "0.2.0"

[dependencies]
biome_console     = { version = "0.0.1", path = "../biome_console" }
biome_diagnostics = { version = "0.2.0", path = "../biome_diagnostics" }
biome_rowan       = { version = "0.2.0", path = "../biome_rowan" }
schemars          = { version = "0.8.10", optional = true }
serde             = { version = "1.0.136", features = ["derive"], optional = true }

[dev-dependencies]
biome_js_factory = { path = "../biome_js_factory" }

[features]
serde = ["dep:serde", "schemars", "biome_rowan/serde"]
