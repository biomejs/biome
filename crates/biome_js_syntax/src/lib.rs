//! A crate for generated Syntax node definitions and utility macros.
//! Both rome_js_lexer and biome_js_parser rely on these definitions, therefore
//! they are wrapped in this crate to prevent cyclic dependencies

#![deny(clippy::use_self)]

#[macro_use]
mod generated;
pub mod assign_ext;
pub mod binary_like_expression;
pub mod binding_ext;
pub mod declaration_ext;
pub mod directive_ext;
pub mod export_ext;
pub mod expr_ext;
pub mod expression_left_side;
pub mod file_source;
pub mod function_ext;
pub mod identifier_ext;
pub mod import_ext;
pub mod jsx_ext;
pub mod misc_ext;
pub mod modifier_ext;
pub mod numbers;
pub mod parameter_ext;
pub mod parentheses;
pub mod static_value;
pub mod stmt_ext;
mod syntax_node;
pub mod type_ext;
mod unescape;
mod union_ext;

pub use self::generated::*;
pub use biome_rowan::{
    SyntaxNodeText, TextLen, TextRange, TextSize, TokenAtOffset, TokenText, TriviaPieceKind,
    WalkEvent,
};
pub use expr_ext::*;
pub use file_source::*;
pub use function_ext::*;
pub use identifier_ext::*;
pub use import_ext::*;
pub use modifier_ext::*;
pub use stmt_ext::*;
pub use syntax_node::*;
pub use unescape::*;

use crate::JsSyntaxKind::*;
use biome_rowan::{AstNode, RawSyntaxKind, SyntaxKind, SyntaxResult};

impl From<u16> for JsSyntaxKind {
    fn from(d: u16) -> Self {
        assert!(d <= (Self::__LAST as u16));
        unsafe { std::mem::transmute::<u16, Self>(d) }
    }
}

impl From<JsSyntaxKind> for u16 {
    fn from(k: JsSyntaxKind) -> Self {
        k as Self
    }
}

impl JsSyntaxKind {
    /// Returns `true` for any contextual (await) or non-contextual keyword
    #[inline]
    pub const fn is_keyword(self) -> bool {
        (self as u16) <= (Self::USING_KW as u16) && (self as u16) >= (Self::BREAK_KW as u16)
    }

    /// Returns `true` for any kind representing a Grit metavariable.
    #[inline]
    pub fn is_metavariable(&self) -> bool {
        matches!(self, Self::GRIT_METAVARIABLE | Self::JS_METAVARIABLE)
    }

    /// Returns `true` for contextual keywords (excluding strict mode contextual keywords)
    #[inline]
    pub const fn is_contextual_keyword(self) -> bool {
        (self as u16) >= (Self::ABSTRACT_KW as u16) && (self as u16) <= (Self::USING_KW as u16)
    }

    /// Returns true for all non-contextual keywords (includes future reserved keywords)
    #[inline]
    pub const fn is_non_contextual_keyword(self) -> bool {
        self.is_keyword() && !self.is_contextual_keyword()
    }

    #[inline]
    pub const fn is_future_reserved_keyword(self) -> bool {
        (self as u16) >= (Self::IMPLEMENTS_KW as u16) && (self as u16) <= (Self::YIELD_KW as u16)
    }
}

impl biome_rowan::SyntaxKind for JsSyntaxKind {
    const TOMBSTONE: Self = TOMBSTONE;
    const EOF: Self = EOF;

    fn is_bogus(&self) -> bool {
        matches!(
            self,
            JS_BOGUS
                | JS_BOGUS_STATEMENT
                | JS_BOGUS_PARAMETER
                | JS_BOGUS_BINDING
                | JS_BOGUS_MEMBER
                | JS_BOGUS_EXPRESSION
                | JS_BOGUS_IMPORT_ASSERTION_ENTRY
                | JS_BOGUS_NAMED_IMPORT_SPECIFIER
                | JS_BOGUS_ASSIGNMENT
                | TS_BOGUS_TYPE
        )
    }

    fn to_bogus(&self) -> Self {
        match self {
            kind if AnyJsModuleItem::can_cast(*kind) => JS_BOGUS_STATEMENT,
            kind if AnyJsExpression::can_cast(*kind) => JS_BOGUS_EXPRESSION,
            kind if AnyJsBinding::can_cast(*kind) => JS_BOGUS_BINDING,
            kind if AnyJsClassMember::can_cast(*kind) || AnyJsObjectMember::can_cast(*kind) => {
                JS_BOGUS_MEMBER
            }
            kind if AnyJsAssignment::can_cast(*kind) => JS_BOGUS_ASSIGNMENT,
            kind if AnyJsNamedImportSpecifier::can_cast(*kind) => JS_BOGUS_NAMED_IMPORT_SPECIFIER,
            kind if AnyJsImportAssertionEntry::can_cast(*kind) => JS_BOGUS_IMPORT_ASSERTION_ENTRY,
            kind if AnyJsParameter::can_cast(*kind) => JS_BOGUS_PARAMETER,
            kind if AnyTsType::can_cast(*kind) => TS_BOGUS_TYPE,

            _ => JS_BOGUS,
        }
    }

    #[inline]
    fn to_raw(&self) -> RawSyntaxKind {
        RawSyntaxKind(*self as u16)
    }

    #[inline]
    fn from_raw(raw: RawSyntaxKind) -> Self {
        Self::from(raw.0)
    }

    fn is_root(&self) -> bool {
        AnyJsRoot::can_cast(*self)
    }

    fn is_list(&self) -> bool {
        Self::is_list(*self)
    }

    fn is_trivia(self) -> bool {
        matches!(
            self,
            Self::NEWLINE | Self::WHITESPACE | Self::COMMENT | Self::MULTILINE_COMMENT
        )
    }

    fn to_string(&self) -> Option<&'static str> {
        Self::to_string(self)
    }
}

impl TryFrom<JsSyntaxKind> for TriviaPieceKind {
    type Error = ();

    fn try_from(value: JsSyntaxKind) -> Result<Self, Self::Error> {
        if value.is_trivia() {
            match value {
                JsSyntaxKind::NEWLINE => Ok(Self::Newline),
                JsSyntaxKind::WHITESPACE => Ok(Self::Whitespace),
                JsSyntaxKind::COMMENT => Ok(Self::SingleLineComment),
                JsSyntaxKind::MULTILINE_COMMENT => Ok(Self::MultiLineComment),
                _ => unreachable!("Not Trivia"),
            }
        } else {
            Err(())
        }
    }
}

/// See: [MDN Operator precedence](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Operator_Precedence#table)
#[derive(Debug, Eq, Ord, PartialOrd, PartialEq, Copy, Clone, Hash)]
pub enum OperatorPrecedence {
    Comma = 0,
    Yield = 1,
    Assignment = 2,
    Conditional = 3,
    Coalesce = 4,
    LogicalOr = 5,
    LogicalAnd = 6,
    BitwiseOr = 7,
    BitwiseXor = 8,
    BitwiseAnd = 9,
    Equality = 10,
    Relational = 11,
    Shift = 12,
    Additive = 13,
    Multiplicative = 14,
    Exponential = 15,
    Unary = 16,
    Update = 17,
    // `new` without arguments list
    NewWithoutArguments = 18,
    LeftHandSide = 19,
    Member = 20,
    Primary = 21,
    Group = 22,
}

impl OperatorPrecedence {
    /// Returns the operator with the lowest precedence
    pub fn lowest() -> Self {
        Self::Comma
    }

    /// Returns the operator with the highest precedence
    pub fn highest() -> Self {
        Self::Primary
    }

    /// Returns `true` if this operator has right to left associativity
    pub fn is_right_to_left(&self) -> bool {
        matches!(
            self,
            Self::Yield | Self::Assignment | Self::Conditional | Self::Exponential | Self::Update
        )
    }

    /// Returns the precedence for a binary operator token or [None] if the token isn't a binary operator
    pub fn try_from_binary_operator(kind: JsSyntaxKind) -> Option<Self> {
        Some(match kind {
            T![??] => Self::Coalesce,
            T![||] => Self::LogicalOr,
            T![&&] => Self::LogicalAnd,
            T![|] => Self::BitwiseOr,
            T![^] => Self::BitwiseXor,
            T![&] => Self::BitwiseAnd,
            T![==] | T![!=] | T![===] | T![!==] => Self::Equality,
            T![<] | T![>] | T![<=] | T![>=] | T![instanceof] | T![in] | T![as] | T![satisfies] => {
                Self::Relational
            }
            T![<<] | T![>>] | T![>>>] => Self::Shift,
            T![+] | T![-] => Self::Additive,
            T![*] | T![/] | T![%] => Self::Multiplicative,
            T![**] => Self::Exponential,
            _ => return None,
        })
    }

    pub const fn is_bitwise(&self) -> bool {
        matches!(self, Self::BitwiseAnd | Self::BitwiseOr | Self::BitwiseXor)
    }

    pub const fn is_shift(&self) -> bool {
        matches!(self, Self::Shift)
    }

    pub const fn is_additive(&self) -> bool {
        matches!(self, Self::Additive)
    }

    pub const fn is_equality(&self) -> bool {
        matches!(self, Self::Equality)
    }

    pub const fn is_multiplicative(&self) -> bool {
        matches!(self, Self::Multiplicative)
    }

    pub const fn is_exponential(&self) -> bool {
        matches!(self, Self::Exponential)
    }
}

/// Similar to [JsSyntaxToken::text_trimmed()], but removes the quotes of string literals.
///
/// ## Examples
///
/// ```
/// use biome_js_syntax::{JsSyntaxKind, JsSyntaxToken, inner_string_text};
///
/// let a = JsSyntaxToken::new_detached(JsSyntaxKind::JS_STRING_LITERAL, "'inner_string_text'", [], []);
/// let b = JsSyntaxToken::new_detached(JsSyntaxKind::JS_STRING_LITERAL, "\"inner_string_text\"", [], []);
/// assert_eq!(inner_string_text(&a), inner_string_text(&b));
///
/// let a = JsSyntaxToken::new_detached(JsSyntaxKind::LET_KW, "let", [], []);
/// let b = JsSyntaxToken::new_detached(JsSyntaxKind::LET_KW, "let", [], []);
/// assert_eq!(inner_string_text(&a), inner_string_text(&b));
///
/// let a = JsSyntaxToken::new_detached(JsSyntaxKind::LET_KW, "let", [], []);
/// let b = JsSyntaxToken::new_detached(JsSyntaxKind::CONST_KW, "const", [], []);
/// assert!(inner_string_text(&a) != inner_string_text(&b));
/// ```
pub fn inner_string_text(token: &JsSyntaxToken) -> TokenText {
    let mut text = token.token_text_trimmed();
    if matches!(
        token.kind(),
        JsSyntaxKind::JS_STRING_LITERAL | JsSyntaxKind::JSX_STRING_LITERAL
    ) {
        // remove string delimiters
        // SAFETY: string literal token have a delimiters at the start and the end of the string
        let range = TextRange::new(1.into(), text.len() - TextSize::from(1));
        text = text.slice(range);
    }
    text
}

/// Returns `Ok(true)` if `maybe_argument` is an argument of a [test call expression](JsCallExpression::is_test_call_expression).
pub fn is_test_call_argument(maybe_argument: &JsSyntaxNode) -> SyntaxResult<bool> {
    let call_expression = maybe_argument
        .parent()
        .and_then(JsCallArgumentList::cast)
        .and_then(|args| args.syntax().grand_parent())
        .and_then(JsCallExpression::cast);

    call_expression.map_or(Ok(false), |call| call.is_test_call_expression())
}
