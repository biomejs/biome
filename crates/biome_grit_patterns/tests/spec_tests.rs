use biome_diagnostics::Diagnostic;
use biome_grit_parser::parse_grit;
use biome_grit_patterns::{
    GritQuery, GritQueryResult, GritTargetFile, GritTargetLanguage, JsTargetLanguage, Message,
    OutputFile,
};
use biome_js_parser::{parse, JsParserOptions};
use biome_js_syntax::JsFileSource;
use biome_test_utils::register_leak_checker;
use grit_util::Range;
use std::{fs::read_to_string, path::Path};

tests_macros::gen_tests! {"tests/specs/**/*.grit", crate::run_test, "module"}

fn run_test(input: &'static str, _: &str, _: &str, _: &str) {
    register_leak_checker();

    let query_path = Path::new(input);

    let (test_name, target_lang_ext) = parse_test_path(query_path);
    if target_lang_ext == "specs" {
        panic!("the test file must be placed in the specs/<target-lang-ext>/ directory");
    }

    // We have one exception: The `specs/error/` directory
    // is for testing error diagnostics.
    if target_lang_ext == "error" {
        run_error_test(query_path, test_name);
        return;
    }

    let Some(target_lang) = GritTargetLanguage::from_extension(target_lang_ext) else {
        panic!("the test file must be placed in the specs/<target-lang-ext>/ directory, unrecognized extension: {target_lang_ext}");
    };

    let query = {
        let query = read_to_string(query_path)
            .unwrap_or_else(|err| panic!("cannot read query from {query_path:?}: {err:?}"));

        let parse_grit_result = parse_grit(&query);
        if !parse_grit_result.diagnostics().is_empty() {
            panic!(
                "cannot parse query from {query_path:?}:\n{:?}",
                parse_grit_result.diagnostics()
            );
        }

        GritQuery::from_node(parse_grit_result.tree(), None, target_lang)
            .unwrap_or_else(|err| panic!("cannot compile query from {query_path:?}: {err:?}"))
    };

    let target_file = {
        let target_path = format!("tests/specs/{target_lang_ext}/{test_name}.{target_lang_ext}");
        let target_path = Path::new(&target_path);
        let target_code = read_to_string(target_path)
            .unwrap_or_else(|err| panic!("failed to read code from {target_path:?}: {err:?}"));
        // TODO: We should generalize this when we have more target languages.
        let target_parse = parse(
            &target_code,
            JsFileSource::tsx(),
            JsParserOptions::default(),
        );

        GritTargetFile {
            path: target_path.into(),
            parse: target_parse.into(),
        }
    };

    let (results, logs) = query
        .execute(target_file)
        .unwrap_or_else(|err| panic!("cannot execute query from {query_path:?}: {err:?}"));
    let snapshot_result = SnapshotResult::from_query_results(results);

    let snapshot = if logs.is_empty() {
        format!("{snapshot_result:#?}")
    } else {
        let logs = logs
            .iter()
            .map(|log| {
                format!(
                    "Message: {}Syntax: {}",
                    log.message,
                    log.syntax_tree.as_deref().unwrap_or_default()
                )
            })
            .collect::<Vec<_>>()
            .join("\n");
        format!("{snapshot_result:#?}\n\n## Logs\n\n{logs}")
    };

    insta::with_settings!({
        prepend_module_to_snapshot => false,
        snapshot_path => query_path.parent().unwrap(),
    }, {
        insta::assert_snapshot!(test_name, snapshot, test_name);
    });
}

fn run_error_test(query_path: &Path, test_name: &str) {
    let snapshot_result = {
        let query = read_to_string(query_path)
            .unwrap_or_else(|err| panic!("cannot read query from {query_path:?}: {err:?}"));

        let parse_grit_result = parse_grit(&query);
        if parse_grit_result.diagnostics().is_empty() {
            match GritQuery::from_node(
                parse_grit_result.tree(),
                None,
                GritTargetLanguage::JsTargetLanguage(JsTargetLanguage),
            ) {
                Ok(_) => panic!("an error was expected when compiling query from {query_path:?}"),
                Err(error) => ErrorSnapshotResult {
                    diagnostics: vec![Box::new(error)],
                },
            }
        } else {
            ErrorSnapshotResult {
                diagnostics: parse_grit_result
                    .diagnostics()
                    .iter()
                    .map(|diagnostic| {
                        let boxed: Box<dyn Diagnostic> = Box::new(diagnostic.clone());
                        boxed
                    })
                    .collect(),
            }
        }
    };

    let snapshot = format!("{snapshot_result:#?}");

    insta::with_settings!({
        prepend_module_to_snapshot => false,
        snapshot_path => query_path.parent().unwrap(),
    }, {
        insta::assert_snapshot!(test_name, snapshot, test_name);
    });
}

/// Tests should be in a `specs/<target-lang-extension>` directory, and each
/// test should have a `.grit` file and a matching `.<target-lang-extension>`
/// file.
///
/// Returns a `("<test-name>", "<target-lang-extension>")` tuple.
fn parse_test_path(file: &Path) -> (&str, &str) {
    let test_name = file.file_stem().unwrap();

    let target_lang_extension = file.parent().unwrap();
    let target_lang_extension = target_lang_extension.file_name().unwrap();

    (
        test_name.to_str().unwrap(),
        target_lang_extension.to_str().unwrap(),
    )
}

#[derive(Debug, Default)]
struct SnapshotResult {
    messages: Vec<Message>,
    matched_ranges: Vec<String>,
    rewritten_files: Vec<OutputFile>,
    created_files: Vec<OutputFile>,
}

impl SnapshotResult {
    fn from_query_results(results: Vec<GritQueryResult>) -> Self {
        let mut snapshot_result = Self::default();
        for result in results {
            match result {
                GritQueryResult::Match(m) => {
                    snapshot_result.messages.extend(m.messages);
                    snapshot_result
                        .matched_ranges
                        .extend(m.ranges.into_iter().map(format_range));
                }
                GritQueryResult::Rewrite(rewrite) => {
                    snapshot_result.messages.extend(rewrite.original.messages);
                    snapshot_result
                        .matched_ranges
                        .extend(rewrite.original.ranges.into_iter().map(format_range));
                    snapshot_result.rewritten_files.push(rewrite.rewritten);
                }
                GritQueryResult::CreateFile(create_file) => {
                    if let Some(ranges) = create_file.ranges {
                        snapshot_result
                            .matched_ranges
                            .extend(ranges.into_iter().map(format_range));
                    }
                    snapshot_result.created_files.push(create_file.rewritten);
                }
            }
        }

        snapshot_result
    }
}

fn format_range(range: Range) -> String {
    format!(
        "{}:{}-{}:{}",
        range.start.line, range.start.column, range.end.line, range.end.column
    )
}

#[derive(Debug, Default)]
struct ErrorSnapshotResult {
    #[allow(unused)]
    diagnostics: Vec<Box<dyn Diagnostic>>,
}
