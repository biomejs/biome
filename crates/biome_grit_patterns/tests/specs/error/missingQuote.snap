---
source: crates/biome_grit_patterns/tests/spec_tests.rs
expression: invalidQuery
---
ErrorSnapshotResult {
    diagnostics: [
        ParseDiagnostic {
            span: Some(
                0..2,
            ),
            message: Missing closing quote,
            advice: ParserAdvice {
                advice_list: [
                    Detail(
                        ParserAdviceDetail {
                            message: "file ends here",
                            span: Some(
                                2..2,
                            ),
                        },
                    ),
                ],
            },
        },
    ],
}
